Simple manual cli tests

holmanb opened the pull request:
<!--
Thank you for submitting a PR to cloud-init!

To ease the process of reviewing your PR, do make sure to complete the following checklist **before** submitting a pull request.

- [ ] I have signed the CLA: https://ubuntu.com/legal/contributors
- [ ] I have added my Github username to ``tools/.github-cla-signers``
- [ ] I have included a comprehensive commit message using the guide below
- [ ] I have added unit tests to cover the new behavior under ``tests/unittests/``
  - Test files should map to source files i.e. a source file ``cloudinit/example.py`` should be tested by ``tests/unittests/test_example.py``
  - Run unit tests with ``tox -e py3``
- [ ] I have kept the change small, avoiding unnecessary whitespace or non-functional changes.
- [ ] I have added a reference to issues that this PR relates to in the PR message (Refs GH-1234, Fixes GH-1234)
- [ ] I have updated the documentation with the changed behavior.
  - If the change doesn't change the user interface and is trivial, this step may be skipped.
  - Cloud-config documentation is generated from the jsonschema.
  - Generate docs with ``tox -e docs``.
-->


## Proposed Commit Message
<!-- Include a proposed commit message because PRs are squash merged
by default.

See https://www.conventionalcommits.org/en/v1.0.0/#specification
for our commit message convention.

If the change is related to a particular cloud or particular distro,
please include the "optional scope" in the summary line. E.g.,
feat(ec2): Add support for foo to the baz

Types used by this project:
feat, fix, docs, ci, test, refactor, chore
-->
Allow calling `cloud-init --all-stages` for testing purposes.
Also add tests for this behavior and various other CLI commands.

## Additional Context
Requires a change in pycloudlib. (update: [merged](https://github.com/canonical/pycloudlib/pull/438))
We could also check that INFO/DEBUG logs don't show up in command output as well - thoughts?

## Test Steps
Check CI integration tests.

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

holmanb committed:
fix: allow manually running --all-stages (#5784)

holmanb committed:
test: support requesting pty from test instance (#5784)

holmanb committed:
test: add fast cli entry point tests (#5784)

These quick-check tests check for the following behaviors on PR:
- return code
- unwanted logs in output

TheRealFalcon reviewed (commented):
Good idea! Over time, I'd prefer the test go away as each of the commands gets their own test verifying expected output / effects rather than having been run without obvious problems. That'll take time though and better to have broad shallow coverage now.

Should this be a draft PR until we can get the pycloudlib piece?

holmanb commented:
> Good idea!

Thanks!

> Over time, I'd prefer the test go away as each of the commands gets their own test verifying expected output / effects rather than having been run without obvious problems. That'll take time though and better to have broad shallow coverage now.

+1, no objections

> Should this be a draft PR until we can get the pycloudlib piece?

Sure, I just put up [the pycloudlib PR](https://github.com/canonical/pycloudlib/pull/438), but I'll drop this to a draft until that merges.

TheRealFalcon reviewed (approved):
LGTM

holmanb committed:
Merge remote-tracking branch 'origin/main' into holmanb/manual-all-stages

holmanb committed:
format

holmanb committed:
fix botched merge

holmanb committed:
drop hanging test

holmanb committed:
drop failing tests

holmanb committed:
reorder collect logs before clean

holmanb committed:
more precise path

holmanb committed:
run systemd test before clean test

holmanb commented:
Changes since previous review:

- merged from upstream
- reordered test to be last in the test class (because cloud-init clean removes logs used in other tests)
- removed full invocations of cloud-init's boot entry points (should be covered by other tests)
- fixed path for `cloud-init collect-logs` (I suspect something changed with pycloudlib?)

blackboxsw commented on tests/integration_tests/modules/test_combined.py:
```diff
@@ -660,6 +681,132 @@ def test_networkd_wait_online(self, class_client: IntegrationInstance):
+            "cloud-init clean --seed",
+            "cloud-init clean --machine-id",
+            "cloud-init clean --configs ssh_config",
+            "cloud-init clean --configs ssh_config",
```
duplicated ssh_config option. Let's use datasource and fstab options instead

blackboxsw commented on tests/integration_tests/modules/test_combined.py:
```diff
@@ -660,6 +681,132 @@ def test_networkd_wait_online(self, class_client: IntegrationInstance):
+                return_code=0,
+            )
+
+    # TODO: add tests for commands as an unpriveledged user
```
typo

```suggestion
    # TODO: add tests for commands as an unprivileged user
```

blackboxsw reviewed (dismissed):
Minor nits. and +1. I'll push changes and land when CI completes. These changes are also tested and via tox -e integration-tests -- tests/integration_tests/modules/test_combined.py

blackboxsw committed:
chore: review comments
